integrations/osmosis-adapter/src/staking.rs

```diff
@@ -300,15 +301,62 @@ pub mod fns {
 
         fn query_rewards(
             &self,
-            _querier: &QuerierWrapper,
+            querier: &QuerierWrapper,
         ) -> Result<RewardTokensResponse, CwStakingError> {
-            Ok(RewardTokensResponse {
-                tokens: Default::default(),
-            })
+            let pool_incentives_querier = PoolincentivesQuerier::new(querier);
+            // let incentives_querier = IncentivesQuerier::new(querier);
+            let tokens = self
+                .tokens
+                .iter()
+                .map(|context| {
+                    pool_incentives_querier
+                        .gauge_ids(context.pool_id)
+                        .map(|gauge_ids_response| {
+                            let assets = gauge_ids_response
+                                .gauge_ids_with_duration
+                                .into_iter()
+                                .map(|gauge_id_with_duration| {
+                                    querier.query::<MiniGaugeByIdResponse>(&osmosis_std::types::osmosis::incentives::GaugeByIdRequest{id: gauge_id_with_duration.gauge_id}.into())
+                                    // TODO: use osmosis api when fixed
+                                    // incentives_querier
+                                    //     .gauge_by_id(gauge_id_with_duration.gauge_id)
+                                        .map(|gauge_by_id_response| {
+                                            gauge_by_id_response
+                                                .gauge
+                                                .unwrap()
+                                                .coins
+                                                .into_iter()
+                                                .map(|coin| AssetInfo::Native(coin.denom))
+                                                .collect::<Vec<_>>()
+                                        })
+                                })
+                                .collect::<StdResult<Vec<_>>>()?;
+                            let mut assets = assets.into_iter().flatten().collect::<Vec<_>>();
+                            assets.dedup();
+                            StdResult::Ok(assets)
+                        })
+                })
+                .collect::<StdResult<StdResult<Vec<_>>>>()??;
+            Ok(RewardTokensResponse { tokens })
         }
     }
 }
 
+#[cfg(feature = "full_integration")]
+// Osmosis returns broken response for `lock_query_type` (Enum stringified instead of Enum.into::<i32> )
+#[derive(cosmwasm_schema::serde::Deserialize)]
+#[serde(rename_all = "snake_case", crate = "::cosmwasm_schema::serde")]
+pub struct MiniGaugeByIdResponse {
+    pub gauge: Option<MiniGauge>,
+}
+
+#[cfg(feature = "full_integration")]
+#[derive(cosmwasm_schema::serde::Deserialize)]
+#[serde(rename_all = "snake_case", crate = "::cosmwasm_schema::serde")]
+pub struct MiniGauge {
+    pub coins: Vec<osmosis_std::types::cosmos::base::v1beta1::Coin>,
```

Comment:
We're really fixing OSMOSIS API because it doesn't work ? Are they planning to update ?

Reply:
Are we sure that this is not only on OsmosisTestube ? Is that working on osmosis-1 or testnet ?

Reply:
> Are we sure that this is not only on OsmosisTestube ? Is that working on osmosis-1 or testnet ?

Not sure if it's test tube bug, but it would work on main or test networks as it ignores unknown fields(and broken field is ignored in this case)

Reply:
Great